### PR TITLE
  Update Working with Dependencies -> Centralizing Dependencies Section

### PR DESCRIPTION
**Pull request by lkasso**
### Details
This is a Documentation change ONLY.

### Context
Create a new Centralizing Dependencies section in [Working with Dependencies](https://docs.gradle.org/current/userguide/declaring_dependencies.html)

#### Centralizing Dependencies:

| Old (Current docs) | New (Build link) | Notes |
| ------------- | ------------- | ------------- |
| [Sharing dependency versions](https://docs.gradle.org/8.9/userguide/platforms.html#sub:using-platform-to-control-transitive-deps) | [Creating Platforms](https://builds.gradle.org/repository/download/Gradle_Release_Check_BuildDistributions/88953610:id/distributions/gradle-8.10.2-docs.zip!/gradle-8.10.2-20240913212843%2B0000/docs/userguide/platforms.html) | Seperated into a dedicated page |
| [Sharing dependency versions](https://docs.gradle.org/8.9/userguide/platforms.html#sub:version-catalog) | [Creating Version Catalogs](https://builds.gradle.org/repository/download/Gradle_Release_Check_BuildDistributions/88953610:id/distributions/gradle-8.10.2-docs.zip!/gradle-8.10.2-20240913212843%2B0000/docs/userguide/version_catalogs.html) | Seperated into a dedicated page |
| None | [Using Catalogs with Platforms](https://builds.gradle.org/repository/download/Gradle_Release_Check_BuildDistributions/88953610:id/distributions/gradle-8.10.2-docs.zip!/gradle-8.10.2-20240913212843%2B0000/docs/userguide/centralizing_catalog_platform.html) | New page |

### Review

Pages to review:

- [platforms.adoc](https://github.com/gradle/gradle/pull/30529/files#diff-67f7a32cd393695a4bd82343e5101c48816e204a1b28d8895ea38ed1ac2fd158)
- [version_catalogs.adoc](https://github.com/gradle/gradle/pull/30529/files#diff-7a78aa6e25f874ae4565bc6437767536d78f6b98c358fe8d180a60b0e51261cd)
- [centralizing_catalog_platform.adoc](https://github.com/gradle/gradle/pull/30529/files#diff-5dc0108d9e0963a88f07fe24fe04d1aad4a0b64f729f334492cc3a0360971892)

### CODE REVIEW

**Comment by lkasso** [comment]
@bot-gradle test BD

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+[[platforms-vs-catalog]]
+= Using Catalogs with Platforms
+
+Both platforms and version catalogs help manage dependency versions in a project, but they serve different purposes and have different effects on dependency resolution:
+
+== Version Catalogs
+
+- **Purpose:** A version catalog is used to centralize and standardize dependency coordinates (group, name, version). It provides type-safe accessors in the build script, making dependencies easier to manage.
+- **Effect on Dependency Graph:** Version catalogs **do not** affect dependency resolution. The versions defined in the catalog are **suggestions** and **transparent to downstream consumers**. These versions can be overridden by other mechanisms like transitive dependencies or platforms.
```
Comment (tresat) [inline platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc:23]: ❗ Can we not say "suggestions" here?

Nothing in a catalog is 1) applied to any configuration by default - the contents need to be explicitly referenced from a `dependencies` block - true, but when they are then 2) they work the same as any "locally" declared dependency.  I don't know how to explain this briefly with one word, but I think it's really important to make this very clear.

I'd also explain that transparent means that the source of a dependency declaration (local or catalog) cannot be identified by downstream consumers.  The declaration itself is not transparent.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+[source,toml]
+.libs.versions.toml
+----
+[libraries]
+mylib = { group = "com.example", name = "mylib", version = "1.0.0" }
+----
+
+== Platforms
+
+- **Purpose:** A platform is a module in the dependency graph that enforces or aligns versions of transitive dependencies. It influences dependency resolution and ensures version consistency across different modules.
```
Comment (tresat) [inline platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc:34]:
```suggestion
- **Purpose:** A platform is a module in the dependency graph that enforces or aligns versions of dependencies (including transitive dependencies). It influences dependency resolution and ensures version consistency across different modules.
```

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+- **Effect on Dependency Graph:** Platforms can **enforce** or **suggest versions** that are propagated through the dependency graph, affecting transitive dependencies and downstream consumers. They are a formal part of the dependency graph and can dictate the version chosen during resolution.
```
Comment (tresat) [inline platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc:35]:
```suggestion
- **Effect on Dependency Graph:** Platforms **apply** or **enforce** versions to dependencies that are declarated locally without versions.  These versions in a platform are propagated through the dependency graph, affecting transitive dependencies and downstream consumers. They are a formal part of the dependency graph and can dictate the version chosen during resolution.
```

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+dependencies {
+    constraints {
+        api("com.example:mylib:2.0.0") // This version will influence resolution
+    }
+}
+----
+
+== Version catalog vs. platform
+
+Even if a version catalog defines a version for a dependency, Gradle might pick a different version during resolution if another component (e.g., a platform or a transitive dependency) suggests a different version.
```
Comment (tresat) [inline platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc:53]: 💭 Unless it's an `enforcedPlatform` - might want to mention/link to that here.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+dependencies {
+    constraints {
+        api("com.example:mylib:2.0.0") // This version will influence resolution
+    }
+}
+----
+
+== Version catalog vs. platform
```
Comment (tresat) [inline platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc:51]: ❗ Don't say vs., as you can use both in the same project - this is mentioned below.  I think both these sections (this and "Using a catalog with a platform") should be combined into just "Using catalogs with platforms")

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+2. **Use platforms** when you need to influence or enforce version alignment across modules. Platforms ensure that dependencies resolve to the desired version, particularly in large or multi-module projects.
+
+== Using a catalog with a platform
+
+A project can use both a version catalog to define dependency versions and a platform to enforce version alignment.
+
+====
+include::sample[dir="snippets/dependencyManagement/catalogs-javaPlatformUsage/kotlin",files="build.gradle.kts[tags=platform_uses_catalog]"]
+include::sample[dir="snippets/dependencyManagement/catalogs-javaPlatformUsage/groovy",files="build.gradle[tags=platform_uses_catalog]"]
+====
```
Comment (tresat) [inline platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc:69]: ❗ This example just shows defining the platform using the catalog.  It should also show a consumer project that uses the platform (perhaps with a different, version-less dependency from the catalog, that itself requires mylib - then the constraint would apply to the transitive myLib dep).

I think this example is too simple, and it is important to more completely demonstrate this scenario.

**Review by tresat — CHANGES_REQUESTED** [review]

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/platforms.adoc [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+This defines a custom platform with specific versions of `guava` and `commons-lang3` that can be applied in other projects.
+
+[[sub:bom_import]]
+== Importing a platform
+
+Gradle supports importing https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Importing_Dependencies[BOMs], which are POM files containing `<dependencyManagement>` sections that manage dependency versions.
+
+=== Regular Platform
+
+To import a BOM, declare it using the `platform` keyword:
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/platforms.adoc:78]: Done.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+[[version-catalog]]
+= Version Catalogs
+
+A version catalog is a curated list of dependencies that can be referenced in build scripts, simplifying dependency management.
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc:18]: Done.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/platforms.adoc [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+====
+include::sample[dir="snippets/dependencyManagement/managingTransitiveDependencies-constraintsFromBOM/kotlin",files="build.gradle.kts[tags=dependency-on-bom]"]
+include::sample[dir="snippets/dependencyManagement/managingTransitiveDependencies-constraintsFromBOM/groovy",files="build.gradle[tags=dependency-on-bom]"]
+====
+
+In this example, the Spring Boot BOM provides the versions for `gson` and `dom4j`, so no explicit versions are needed.
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/platforms.adoc:85]: Added.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/platforms.adoc [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+=== Enforced Platform
+
+The `enforcedPlatform` keyword can be used to override any versions found in the dependency graph, but should be used with caution as it affects all consumers of your project:
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/platforms.adoc:89]: Updated and added back the disclaimer.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+
+----
+a.b = { c = "d" }
+----
+
+or
+
+----
+a = { b = { c = "d" } }
+----
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc:88]: Removed.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+[libraries]
+my-lib = { group = "com.mycompany", name="mylib", version.ref="some" }
+----
+
+[[sec:dependency-bundles]]
+==== Bundles
+
+Bundles group multiple library aliases, so they can be referenced together in the build script.
+
+[source,toml]
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc:144]: Done.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+----
+[plugins]
+versions = { id = "com.github.ben-manes.versions", version = "0.45.0" }
+----
+
+Which can be accessed in any project of the build using the `plugins{}` block:
+
+====
+include::sample[dir="snippets/dependencyManagement/catalogs-settings/kotlin",files="build.gradle.kts[tags=use_plugin]"]
+include::sample[dir="snippets/dependencyManagement/catalogs-settings/groovy",files="build.gradle[tags=use_plugin]"]
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc:173]: Done.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+====
+
+[[sub:type-unsafe-access-to-catalog]]
+Accessors map to the aliases and versions defined in the TOML file, providing type-safe access to dependencies and plugins.
+
+[[sub:mapping-aliases-to-accessors]]
+=== Aliases and type-safe accessors
+
+Aliases in a version catalog consist of identifiers separated by a dash (`-`), underscore (`_`), or dot (`.`).
+Type-safe accessors are generated for each alias, normalized to dot notation:
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc:227]: Done.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+However, if you need to import version catalogs from multiple files, it's recommended to use a code-based approach instead of relying on TOML files.
+This approach allows for the declaration of multiple catalogs from different files:
+
+====
+include::sample[dir="snippets/dependencyManagement/catalogs-toml/kotlin",files="settings.gradle.kts[tags=additional_catalog]"]
+include::sample[dir="snippets/dependencyManagement/catalogs-toml/groovy",files="settings.gradle[tags=additional_catalog]"]
+====
+
+[[sec:multiple]]
+== Using multiple catalogs
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc:318]: I renamed the section.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+====
+
+In the examples above, any dependency referencing the `groovy` version will automatically be updated to use `3.0.6`.
+
+NOTE: Overwriting a version only affects what is imported and used when declaring dependencies. The actual resolved dependency version may differ due to conflict resolution.gi
+
+[[sec:programmatic-catalog-versions]]
+== Programming catalogs
+
+Version catalogs can be declared programmatically in the `settings.gradle(.kts)` file.
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/version_catalogs.adoc:378]: Done.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc (hunk 1 above)
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc:23]: Done. Might need a final check.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc (hunk 2 above)
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc:34]: Done.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc (hunk 3 above)
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc:35]: Done.

File: platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc (hunk 6 above)
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/centralizing-dependencies/centralizing_catalog_platform.adoc]: Done. Added a brand new snippet.

**Comment by lkasso** [comment]
@bot-gradle test BD

**Comment by lkasso** [comment]
@bot-gradle clean

**Review by tresat — APPROVED** [review]

**Pull request merged**